syntax = "proto3";

package sintellix;

import "model_state.proto";

// CIC (Channel-in-Channel) Data Structure
// Supports multi-channel data packaging and extraction
message CICData {
  // Individual channel within CIC
  message Channel {
    string name = 1;                    // Channel name (e.g., "text", "image", "audio")
    ChannelType type = 2;               // Channel type
    bytes data = 3;                     // Channel data (serialized embedding or raw data)
    uint32 dimension = 4;               // Data dimension
    map<string, string> metadata = 5;   // Additional metadata
  }

  // Channel types
  enum ChannelType {
    CHANNEL_TYPE_UNSPECIFIED = 0;
    CHANNEL_TYPE_TEXT = 1;              // Text embedding channel
    CHANNEL_TYPE_IMAGE = 2;             // Image embedding channel
    CHANNEL_TYPE_AUDIO = 3;             // Audio embedding channel
    CHANNEL_TYPE_COMPOSITE = 4;         // Composite channel (multiple channels merged)
    CHANNEL_TYPE_RAW = 5;               // Raw data channel
  }

  // All channels in this CIC
  repeated Channel channels = 1;

  // CIC metadata
  string cic_id = 2;                    // Unique CIC identifier
  uint64 timestamp = 3;                 // Creation timestamp
  map<string, string> metadata = 4;     // Additional metadata
}

// CIC extraction request
message CICExtractionRequest {
  CICData source_cic = 1;               // Source CIC data
  repeated string channel_names = 2;    // Channels to extract
}

// CIC extraction response
message CICExtractionResponse {
  repeated CICData.Channel channels = 1; // Extracted channels
  bool success = 2;
  string error_message = 3;
}

// CIC repackaging request
message CICRepackagingRequest {
  repeated CICData.Channel channels = 1; // Channels to repackage
  string composite_name = 2;             // Name for composite channel
  map<string, string> metadata = 3;      // Additional metadata
}

// CIC repackaging response
message CICRepackagingResponse {
  CICData.Channel composite_channel = 1; // Repackaged composite channel
  bool success = 2;
  string error_message = 3;
}
